staging/operator-lifecycle-manager/pkg/controller/operators/packageserver_controller.go

```diff
@@ -0,0 +1,165 @@
+	// TODO(tflannag): Should this be a custom struct instead?
+	// TODO(tflannag): Add a mutex here?
+	// TODO(tflannag): Add an event recorder?
+	HighlyAvailableMode bool
+}
+
+//+kubebuilder:rbac:groups=config.openshift.io,resources=infrastructures,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=config.openshift.io,resources=infrastructures/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=config.openshift.io,resources=infrastructures/finalizers,verbs=update
+func (r *PackageServerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("deployment", req.NamespacedName)
+
+	log.Info("handling current request", "request", req.String())
+	log.Info("currently topology mode", "highly available", r.HighlyAvailableMode)
+	defer log.Info("finished request reconciliation")
+
+	deployment := &appsv1.Deployment{}
+	err := r.Client.Get(ctx, req.NamespacedName, deployment)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+	if apierrors.IsNotFound(err) {
+		deployment, err := r.syncDeployment()
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, r.Client.Create(ctx, deployment)
+	}
+
+	// TODO(tflannag): Need to handle updating any non-replica high availability fields
+	// e.g. hard anti affinity, maxUnavailable, etc.
+	expectedReplicas := int32(defaultReplicaCount)
+	if !r.HighlyAvailableMode {
+		expectedReplicas = int32(singleReplicaCount)
+	}
+	if *deployment.Spec.Replicas == expectedReplicas {
+		log.Info("deployment does not require any updates")
+		return ctrl.Result{}, nil
+	}
+
+	// TODO(tflannag): May want to avoid using Update and instead migrate towards
+	// using SSA patching. Can also just nuke any spec fields that don't match
+	// what's being defined in the default deployment.yaml.
+	log.Info("updating replica count", "old replicas", *deployment.Spec.Replicas, "new replicas", expectedReplicas)
+	deployment.Spec.Replicas = pointer.Int32Ptr(expectedReplicas)
+	if err := r.Client.Update(ctx, deployment); err != nil {
+		log.Error(err, "failed to update the packageserver deployment")
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PackageServerReconciler) syncDeployment() (*appsv1.Deployment, error) {
+	deployment, err := manifests.NewPackageServerDeployment(r.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("failed to serialize deployment manifest into Go structure: %v", err)
+	}
+
+	if !r.HighlyAvailableMode {
+		r.Log.Info("packageserver will be deployed in non-HA mode")
+		deployment.Spec.Replicas = pointer.Int32Ptr(1)
+	}
+
+	return deployment, nil
+}
+
+func (r *PackageServerReconciler) infrastructureHandler(obj client.Object) []reconcile.Request {
+	// TODO(tflannag): Is this thread safe? Do I need to instantiate a mutex in the reconciler struct?
+	// TODO(tflannag): Is this an abuse of the mapping handler function? Should we be altering state here?
```

Comment:
If we miss an event on the Infrastructure resource, the controller can be easily left in an inconsistent state, which can lead to incorrect scaling when the packageserver's replicas are set.

I suggest making a client/cache call in the `Reconcile` function instead.